### PR TITLE
Add JSON.MERGE

### PR DESCRIPTION
**Pull request by gkorland**

### CODE REVIEW

File: src/serde_value_manager.rs [hunk 1]
```diff
@@ -222,14 +222,30 @@ impl<'a> WriteHolder<Value, Value> for KeyHolderWrite<'a> {
...
-            update(&path, self.get_value().unwrap().unwrap(), |_v| {
+            replace(&path, self.get_value()?.unwrap(), |_v| {
                 updated = true;
                 Ok(Some(v.take()))
             })?;
         }
         Ok(updated)
     }
 
+    fn merge_value(&mut self, path: Vec<String>, v: Value) -> Result<bool, RedisError> {
```
Comment (oshadmi) [inline src/serde_value_manager.rs:233]: Not using `json-patch` crate (Ref #150) since it only works with serde_json `Value` and not wiith `IValue`?

File: tests/pytest/test.py [hunk 2]
```diff
@@ -1279,6 +1279,15 @@ def testFilterPrecedence(env):
     r.assertEqual(json.loads(res), [doc[0]])
     r.expect('JSON.GET', 'doc', '$[?(@.f==true || @.one==1 && @.t==false)]').equal('[]')
 
+def testMerge(env):
```
Comment (oshadmi) [inline tests/pytest/test.py:1282]: Need test for key deletion? (merge with `null`)

File: src/ivalue_manager.rs [hunk 3]
```diff
@@ -569,6 +585,25 @@ impl ReadHolder<IValue> for IValueKeyHolderRead {
     }
 }
 
+fn merge(doc: &mut IValue, patch: &IValue) {
+    if !patch.is_object() {
+        *doc = patch.clone();
+        return;
+    }
+
+    if !doc.is_object() {
```
Comment (oshadmi) [inline src/ivalue_manager.rs:594]: If doc is not an object, it will be completely overwritten (replaced) by the patch value? (merge takes all from the right-hand-side patch object, and nothing from the left-hand-side array or scalar)

**Review by oshadmi — COMMENTED** [review]
Few comments

File: src/serde_value_manager.rs (hunk 1 above)
Comment (gkorland) [reply src/serde_value_manager.rs:233]: right, this is why I had to implement merge myself

File: src/commands.rs [hunk 4]
```diff
@@ -665,6 +665,85 @@ pub fn json_set<M: Manager>(manager: M, ctx: &Context, args: Vec<RedisString>) -
...
+    let mut redis_key = manager.open_key_write(ctx, key)?;
+    let current = redis_key.get_value()?;
+
+    let val = manager.from_str(value, format)?;
+
+    match current {
+        Some(doc) => {
+            if path.get_path() == JSON_ROOT_PATH {
+                redis_key.merge_value(Vec::new(), val)?;
+                redis_key.apply_changes(ctx, "json.merge")?;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/commands.rs:698]: https://github.com/RediSearch/RediSearch/pull/3342

File: src/ivalue_manager.rs (hunk 3 above)
Comment (gkorland) [reply src/ivalue_manager.rs:594]: yes, I didn't do the arrays support yet

**Comment by joshuakoh1** [comment]
@gkorland Can we also add this "merge" functionality to numincrby where it treats the original value as 0 if the key does not exist?

File: src/ivalue_manager.rs (hunk 3 above)
Comment (gkorland) [reply src/ivalue_manager.rs:576]: I think this is what we agreed on merge arrays, right? @adrianoamaral ?

**Comment by gkorland** [comment]
> functionality

I think it worth a different feature 
@adrianoamaral

File: src/ivalue_manager.rs (hunk 3 above)
Comment (adrianoamaral) [reply src/ivalue_manager.rs:576]: yes, following the RFC7396 - array, replaced by the new value. Other behaviours we can handle on the JSON.ARRAPPEND

**Comment by adrianoamaral** [comment]
Already on the plans for enhance JSON.NUMINCRBY @joshuakoh1, thank you for asking.

File: tests/pytest/test.py
Comment (oshadmi) [inline tests/pytest/test.py:1]: Need to also test key notifications? As with [`JSON.MSET`](https://github.com/RedisJSON/RedisJSON/blob/8202665ae20e00a93be9aa6ecf6021cdd0c5bf5a/tests/pytest/test_keyspace_notifications.py#LL33C56-L33C56)

Comment (gkorland) [reply tests/pytest/test.py:1]: done

**Review by oshadmi — APPROVED** [review]
